analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs

```diff
@@ -55,12 +58,25 @@ protected SonarAnalysisContextBase(SonarAnalysisContext analysisContext, TContex
 
     /// <param name="tree">Tree to decide on. Can be null for Symbol-based and Compilation-based scenarios. And we want to analyze those too.</param>
     /// <param name="generatedCodeRecognizer">When set, generated trees are analyzed only when language-specific 'analyzeGeneratedCode' configuration property is also set.</param>
-    public bool ShouldAnalyzeTree(SyntaxTree tree, GeneratedCodeRecognizer generatedCodeRecognizer) =>
-        SonarLintXml() is var sonarLintXml
-        && (generatedCodeRecognizer is null
-            || sonarLintXml.AnalyzeGeneratedCode(Compilation.Language)
-            || !tree.IsConsideredGenerated(generatedCodeRecognizer, Compilation, IsRazorAnalysisEnabled()))
-        && (tree is null || (!IsUnchanged(tree) && !IsExcluded(sonarLintXml, tree.FilePath)));
+    public bool ShouldAnalyzeTree(SyntaxTree tree, GeneratedCodeRecognizer generatedCodeRecognizer)
```

Comment:
Can you open a second PR with this change in `SonarCompilationStartAnalysisContext`
```cs
    public void RegisterNodeAction<TSyntaxKind>(GeneratedCodeRecognizer generatedCodeRecognizer, Action<SonarSyntaxNodeReportingContext> action, params TSyntaxKind[] syntaxKinds)
        where TSyntaxKind : struct
    {
        if (HasMatchingScope(AnalysisContext.SupportedDiagnostics))
        {
            var lastShouldAnalyze = default(Tuple<SyntaxTree, bool>);
            Context.RegisterSyntaxNodeAction(x =>
            {
                var tree = x.Node.SyntaxTree;
                var last = lastShouldAnalyze; // Make a local copy of the reference to avoid concurrency issues between the access of Item1 and Item2
                bool shouldAnalyze;
                if (ReferenceEquals(last?.Item1, tree))
                {
                    shouldAnalyze = last!.Item2;
                }
                else
                {
                    // Inlined from "Execute"
                    shouldAnalyze = ShouldAnalyzeTree(x.Node.SyntaxTree, generatedCodeRecognizer)
                        && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(AnalysisContext.SupportedDiagnostics, x.Node.SyntaxTree)
                        && AnalysisContext.ShouldAnalyzeRazorFile(x.Node.SyntaxTree);
                    lastShouldAnalyze = new Tuple<SyntaxTree, bool>(tree, shouldAnalyze);
                }
                if (shouldAnalyze)
                {
                    action(new(AnalysisContext, x));
                }
            }, syntaxKinds);
        }
    }
```

It would be interesting to see the savings here, as this doesn't come with the overhead of the concurrent dictionaries.

Reply:
Using this approach is not as good as caching in a dictionary.
![image](https://github.com/SonarSource/sonar-dotnet/assets/126669183/b8049aba-c789-4e59-b1ef-598e3105d39b)
But that's only from `ShouldAnalyzeTree` point of view.

From `SonarCompilationStartAnalysisContext.RegisterNodeAction` point of view, it is faster:

Using your solution, with a tuple:
![image](https://github.com/SonarSource/sonar-dotnet/assets/126669183/d50fa9c3-52e6-448e-85e2-e6a8662a6aae)
My solution, with a ConcurrentDisctionary:
![image](https://github.com/SonarSource/sonar-dotnet/assets/126669183/f7336031-2f84-4e6c-a77d-afa7425aed3b)

For reference, here it is before any changes (of this PR):
![image](https://github.com/SonarSource/sonar-dotnet/assets/126669183/db9dc600-3995-4682-ab5c-fa06c2f10559)

Reply:
Am I right in assuming that this is guaranteed that within the same compilation, the SyntaxTrees will be analyzed sequentially?
Once `lastShouldAnalyze` will change that means analysis on the last SyntaxTree is done?

Reply:
Please open two (draft) PRs so we can compare both solutions.

> Am I right in assuming that this is guaranteed that within the same compilation, the SyntaxTrees will be analyzed sequentially?

There is no such guarantee to rely on:

> There are no ordering constraints for the invocations of any actions, except those introduced by start and end actions (compilation start, compilation end, symbol start, symbol end, operation block start, operation block end, code bock start, and code block end actions). Invocations of actions of one kind may be interleaved with invocations of actions of another kind. A given host environment might appear to provide a predictable order—for example, invoking syntax tree actions before semantic model actions, or symbol actions before code block actions, but depending on such an ordering is not safe.

[Source](https://github.com/dotnet/roslyn/blob/main/docs/analyzers/Analyzer%20Actions%20Semantics.md)

It was a shot in the blue by me. We can make an educated guess now for the analyzer runner that is in place in the build context. That runner is likely different from the one running in the IDE context. Your solution is more robust but comes at the cost of:
* Locking on write (which is somewhat common)
* Slow on a large number of SyntaxTree per compilation
* Holds references to all SyntaxTrees at the end of the compilation, even though only a small portion is still needed
* Does not work well when one of the additional files changes (namely SonarLint.xml).

We may need to come up with a compromise between only covering "latest" (mine) and "all" (yours). With the following characteristics
* Cache is limited in size (e.g. up CPU cores elements, or up to ~12 elements (linear search for up to 12 elements is considered as fast as a hash-based lookup, see e.g. https://terrainformatica.com/2017/10/15/when-linear-search-is-faster-than-stdmapfind-and-stdunordered_mapfind/))
* First in first out or last accessed semantics
* Lock-free at the cost of imprecision

At the same time, we do not have the time to over-engineer it.

Reply:
I will keep this one PR for the tuple solution.
I opened #8411 with the `ConcurrentDirectory` solution